Importing all necessary packages to manipulate the data

In [1]:
import numpy as np
import pandas as pd
import pymysql
import sqlalchemy

In [2]:
originalData=pd.read_csv('./CrimeData.csv')
originalData.tail(5)

incident_id        date      state city_or_county  \
239672      1083142  2018-03-31  Louisiana          Rayne   
239673      1083139  2018-03-31  Louisiana   Natchitoches   
239674      1083151  2018-03-31  Louisiana         Gretna   
239675      1082514  2018-03-31      Texas        Houston   
239676      1081940  2018-03-31      Maine   Norridgewock   

                                   address  n_killed  n_injured  \
239672  North Riceland Road and Highway 90         0          0   
239673                      247 Keyser Ave         1          0   
239674           1300 block of Cook Street         0          1   
239675              12630 Ashford Point Dr         1          0   
239676                    434 Skowhegan Rd         2          0   

                                             incident_url  \
239672  http://www.gunviolencearchive.org/incident/108...   
239673  http://www.gunviolencearchive.org/incident/108...   
239674  http://www.gunviolencearchive.org/incident/108...   
239675  http://www.gunviolencearchive.org/incident/108...   
239676  http://www.gunviolencearchive.org/incident/108...   

                                               source_url  \
239672  http://www.klfy.com/news/local/rayne-woman-cha...   
239673  http://www.ksla.com/story/37854648/man-wanted-...   
239674  http://www.nola.com/crime/index.ssf/2018/04/sh...   
239675  https://www.chron.com/news/houston-texas/houst...   
239676  https://www.centralmaine.com/2018/03/31/police...   

        incident_url_fields_missing          ...           participant_age  \
239672                        False          ...                     0::25   
239673                        False          ...                     1::21   
239674                        False          ...                     0::21   
239675                        False          ...                     0::42   
239676                        False          ...              0::58||1::62   

             participant_age_group  participant_gender  \
239672                0::Adult 18+           0::Female   
239673  0::Adult 18+||1::Adult 18+    0::Male||1::Male   
239674                0::Adult 18+             0::Male   
239675                0::Adult 18+             0::Male   
239676  0::Adult 18+||1::Adult 18+  0::Female||1::Male   

                                      participant_name  \
239672                                0::Jhkeya Tezeno   
239673  0::Jamal Haskett||1::Jaquarious Tyjuan Ardison   
239674                                             NaN   
239675                                  0::Leroy Ellis   
239676        0::Marie Lancaster Hale||1::William Hale   

                         participant_relationship  \
239672                                        NaN   
239673                                        NaN   
239674                                        NaN   
239675                                        NaN   
239676  1::Significant others - current or former   

                      participant_status               participant_type  \
239672             0::Unharmed, Arrested             0::Subject-Suspect   
239673  0::Killed||1::Unharmed, Arrested  0::Victim||1::Subject-Suspect   
239674                        0::Injured                      0::Victim   
239675                         0::Killed                      0::Victim   
239676              0::Killed||1::Killed  0::Victim||1::Subject-Suspect   

                                                  sources  \
239672  http://www.klfy.com/news/local/rayne-woman-cha...   
239673  http://www.ksla.com/story/37854648/man-wanted-...   
239674  http://www.nola.com/crime/index.ssf/2018/04/sh...   
239675  http://www.khou.com/article/news/hpd-investiga...   
239676  https://www.centralmaine.com/2018/03/31/police...   

       state_house_district state_senate_district  
239672                  NaN                   NaN  
239673                 23.0                  31.0  
239674                 85.0                 

In [3]:
originalData[['congressional_district', 'gun_stolen',
       'gun_type', 'incident_characteristics', 'latitude',
       'location_description', 'longitude', 'n_guns_involved', 'notes',]].tail(50)

congressional_district              gun_stolen  \
239627                     4.0              0::Unknown   
239628                     4.0              0::Unknown   
239629                     1.0              0::Unknown   
239630                    13.0              0::Unknown   
239631                     NaN              0::Unknown   
239632                    12.0              0::Unknown   
239633                     NaN              0::Unknown   
239634                    29.0              0::Unknown   
239635                     2.0              0::Unknown   
239636                     7.0              0::Unknown   
239637                     NaN              0::Unknown   
239638                    21.0              0::Unknown   
239639                    20.0              0::Unknown   
239640                     7.0              0::Unknown   
239641                     8.0              0::Unknown   
239642                     4.0  0::Unknown||1::Unknown   
239643                     2.0              0::Unknown   
239644                     NaN              0::Unknown   
239645                     9.0              0::Unknown   
239646                    14.0              0::Unknown   
239647                     NaN              0::Unknown   
239648                     NaN              0::Unknown   
239649                     NaN              0::Unknown   
239650                    25.0              0::Unknown   
239651                    13.0              0::Unknown   
239652                    30.0              0::Unknown   
239653                     2.0              0::Unknown   
239654                     2.0              0::Unknown   
239655                    52.0              0::Unknown   
239656                    25.0              0::Unknown   
239657                    21.0              0::Unknown   
239658                     NaN              0::Unknown   
239659                    10.0              0::Unknown   
239660                     9.0              0::Unknown   
239661                     NaN              0::Unknown   
239662                     NaN              0::Unknown   
239663                     3.0              0::Unknown   
239664                     8.0              0::Unknown   
239665                     NaN              0::Unknown   
239666                     NaN              0::Unknown   
239667                     9.0              0::Unknown   
239668                     NaN              0::Unknown   
239669                     NaN              0::Unknown   
239670                     NaN              0::Unknown   
239671                     5.0              0::Unknown   
239672                     NaN              0::Unknown   
239673                     4.0              0::Unknown   
239674                     2.0              0::Unknown   
239675                     9.0              0::Unknown   
239676                     2.0  0::Unknown||1::Unknown   

                      gun_type  \
239627              0::Unknown   
239628              0::Unknown   
239629              0::Unknown   
239630              0::Unknown   
239631              0::Unknown   
239632              0::Unknown   
239633              0::Unknown   
239634              0::Unknown   
239635              0::Unknown   
239636              0::Unknown   
239637              0::Unknown   
239638              0::Unknown   
239639              0::Unknown   
239640              0::Unknown   
239641              0::Unknown   
239642  0::Handgun||1::Handgun   
239643              0::Unknown   
239644              0::Unknown   
239645              0::Unknown   
239646              0::Unknown   
239647              0::Unknown   
239648              0::Unknown   
239649              0::Unknown   
239650              0::Unknown   
239651              0::Unknown   
239652              0::Unknown   
239653              0::Unknown   
239654              0::Unknown   
239655              0::Unknown   
239656              0::Unknow

In [4]:
originalData.columns

Index(['incident_id', 'date', 'state', 'city_or_county', 'address', 'n_killed',
       'n_injured', 'incident_url', 'source_url',
       'incident_url_fields_missing', 'congressional_district', 'gun_stolen',
       'gun_type', 'incident_characteristics', 'latitude',
       'location_description', 'longitude', 'n_guns_involved', 'notes',
       'participant_age', 'participant_age_group', 'participant_gender',
       'participant_name', 'participant_relationship', 'participant_status',
       'participant_type', 'sources', 'state_house_district',
       'state_senate_district'],
      dtype='object')

In [5]:
originalData.describe()

incident_id       n_killed      n_injured  congressional_district  \
count  2.396770e+05  239677.000000  239677.000000           227733.000000   
mean   5.593343e+05       0.252290       0.494007                8.001265   
std    2.931287e+05       0.521779       0.729952                8.480835   
min    9.211400e+04       0.000000       0.000000                0.000000   
25%    3.085450e+05       0.000000       0.000000                2.000000   
50%    5.435870e+05       0.000000       0.000000                5.000000   
75%    8.172280e+05       0.000000       1.000000               10.000000   
max    1.083472e+06      50.000000      53.000000               53.000000   

            latitude      longitude  n_guns_involved  state_house_district  \
count  231754.000000  231754.000000    140226.000000         200905.000000   
mean       37.546598     -89.338348         1.372442             55.447132   
std         5.130763      14.359546         4.678202             42.048117   
min        19.111400    -171.429000         1.000000              1.000000   
25%        33.903400     -94.158725         1.000000             21.000000   
50%        38.570600     -86.249600         1.000000             47.000000   
75%        41.437375     -80.048625         1.000000             84.000000   
max        71.336800      97.433100       400.000000            901.000000   

       state_senate_district  
count           207342.00000  
mean                20.47711  
std                 14.20456  
min                  1.00000  
25%                  9.00000  
50%                 19.00000  
75%                 30.00000  
max                 94.00000

In [6]:
len(originalData.incident_id)

239677

The total number of crimes in USA where a fire gun was involved from 2013 to 2018 is 239677. In this study we will look into more detail in the numbers to discover the nature and chronologic evolution of crime.

Some of the columns in the data are not relevant for what we need to check so we can get rid of a few columns in  order to have a more handleable dataset

In [7]:
workingData=originalData.drop(['address','gun_stolen', 'gun_type','incident_characteristics','notes','incident_url','participant_age','participant_gender','participant_status','participant_type', 'source_url',
       'incident_url_fields_missing', 'congressional_district', 'latitude',
       'location_description', 'longitude',
       'participant_age_group', 'participant_name',
       'sources', 'state_house_district',
       'state_senate_district'], axis=1)
workingData.head(7)

incident_id        date           state city_or_county  n_killed  \
0       461105  2013-01-01    Pennsylvania     Mckeesport         0   
1       460726  2013-01-01      California      Hawthorne         1   
2       478855  2013-01-01            Ohio         Lorain         1   
3       478925  2013-01-05        Colorado         Aurora         4   
4       478959  2013-01-07  North Carolina     Greensboro         2   
5       478948  2013-01-07        Oklahoma          Tulsa         4   
6       479363  2013-01-19      New Mexico    Albuquerque         5   

   n_injured  n_guns_involved participant_relationship  
0          4              NaN                      NaN  
1          3              NaN                      NaN  
2          3              2.0                      NaN  
3          0              NaN                      NaN  
4          2              2.0                3::Family  
5          0              NaN                      NaN  
6          0              2.0                5::Family

We now need to identify the null values in our dataset, the best way to this is calculating the number of nulls per column and calculating their percentage over the total amount of registers.

In [8]:
null_cols = workingData.isnull().sum()*100/len(workingData.incident_id)
null_cols[null_cols > 0]

n_guns_involved             41.493760
participant_relationship    93.418643
dtype: float64

We can observe that the column "participant_relationship" contains approximately 93% of null values. Due to the nature of the information contained in that column, we can assume that no value means no relationship. Therefore, we will exchange the nulls with the string "False". In the same manner, we are not really interested in the kind of relationship, but in the existance of a relationship, therefore, registers with a value can be exchanged with "True"

In [9]:
# We are fillin NA 
workingData['participant_relationship']=workingData.participant_relationship.fillna(0)
workingData['participant_relationship'].head()

0            0
1            0
2            0
3            0
4    3::Family
Name: participant_relationship, dtype: object

In [10]:
workingData.participant_relationship[workingData.participant_relationship!=0] = 1
workingData.participant_relationship.head()

/home/albertocastellanos/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


0    0
1    0
2    0
3    0
4    1
Name: participant_relationship, dtype: object

From checking the data types, we know that some of the columns are expressed with the wrong data type so we should change it.

In [11]:
workingData.dtypes

incident_id                   int64
date                         object
state                        object
city_or_county               object
n_killed                      int64
n_injured                     int64
n_guns_involved             float64
participant_relationship     object
dtype: object

In [12]:
workingData['participant_relationship']=workingData.participant_relationship.astype('bool')
workingData['participant_relationship'].head(5)

0    False
1    False
2    False
3    False
4     True
Name: participant_relationship, dtype: bool

In [13]:
workingData.date=workingData.date.astype('datetime64[ns]')
workingData.date.head(5)

0   2013-01-01
1   2013-01-01
2   2013-01-01
3   2013-01-05
4   2013-01-07
Name: date, dtype: datetime64[ns]

We know that "n-guns-involved"'s NaN values never equal zero since the current dataset refers to crimes where a fire gun was involved. NaN values simply means that the police was unable to clarify the number of guns present in the crime, so this value could vary quite a lot. Therefore, we have decided that we just want to work with cases where we know the number of guns that were used in the crime. 

In [14]:
workingData=workingData.dropna(subset=['n_guns_involved'])
workingData.head()

incident_id       date           state city_or_county  n_killed  \
2        478855 2013-01-01            Ohio         Lorain         1   
4        478959 2013-01-07  North Carolina     Greensboro         2   
6        479363 2013-01-19      New Mexico    Albuquerque         5   
10       491674 2013-01-23       Tennessee    Chattanooga         1   
11       479413 2013-01-25        Missouri    Saint Louis         1   

    n_injured  n_guns_involved  participant_relationship  
2           3              2.0                     False  
4           2              2.0                      True  
6           0              2.0                      True  
10          3              1.0                     False  
11          3              1.0                     False

In [15]:
workingData.isnull().sum()

incident_id                 0
date                        0
state                       0
city_or_county              0
n_killed                    0
n_injured                   0
n_guns_involved             0
participant_relationship    0
dtype: int64

We can now be sure that there are not null values in our dataset and therefore we are safe to extract conclusions from it

The name of the columns was not as descriptive as desired so we have decided to rename them as necessary:

In [16]:
workingData.columns

Index(['incident_id', 'date', 'state', 'city_or_county', 'n_killed',
       'n_injured', 'n_guns_involved', 'participant_relationship'],
      dtype='object')

In [17]:
workingData=workingData.rename(columns={'city_or_county':'city'})

In [18]:
workingData.columns

Index(['incident_id', 'date', 'state', 'city', 'n_killed', 'n_injured',
       'n_guns_involved', 'participant_relationship'],
      dtype='object')

In [19]:
workingData.columns = ['crimeId', 'date', 'state', 'city', 'numberKilled',
       'numberInjured', ' numberGunsInvolved', 'authorsRelationship']

We have also decided to re-arrange the column order for clarity:

In [20]:
columnOrder = ['crimeId', 'date', 'state', 'city', 'authorsRelationship', 'numberKilled',
       'numberInjured', ' numberGunsInvolved']

Data is ready to be analized so we can now export it to .csv now:

In [22]:
workingData.to_csv('./CrimeData_Cleaned.csv', index=False)

Ranking of 10 states whith higher number of crimes

In [ ]:
workingData['state'].value_counts().head(10)

Ranking of 10 states by number of crimes when killed=1

In [ ]:
Top10for5killed=workingData[(workingData.numberKilled==1)]
Top10for5killed['state'].value_counts().head(10)

In [ ]:
workingData.head(5)

We now want to discover the 10 days where more crimes with victims but no casulties happened in the city of Chicago

In [ ]:
dayRankVictims=workingData[(workingData.city=="Chicago") & (workingData.numberInjured>=1) & (workingData.numberKilled==0)]
dayRankVictims.date.value_counts().head(10)

We are going to categorize each register based on the number of killed people in the event. For that purpose we will use bin in order to establish 4 groups.

In [ ]:
numberOfKills=["Zero to Three", "Four to Six", "Six to Nine", "Ten or more"]
cutoffs=[-1,3,6,9,999]
bins=pd.cut(workingData["numberKilled"],cutoffs,labels=numberOfKills)
print(workingData.columns)
dfbins=bins.to_frame()
dfbins.columns=['bins']
dfbins.bins.value_counts()




We can observe that the number of mass killings is 6, we want now to discover the states where more mass killings happened:

In [ ]:
dataWithBins=pd.concat([workingData, dfbins], axis=1)
dataWithBins

I have been trying to resolve a group by with several columns that aggregates a column and sort by it but I have been unable to achieve any result

dataWithBins.groupby(['bins', 'state']).count().sort_values("numberKilled")

In [ ]:
dataWithBins.loc[dataWithBins['bins'].str.startswith('Ten'), 'Mass Killing']=True
dataWithBins.loc[(dataWithBins['bins'].str.startswith('Ten')==False), 'Mass Killing']=False

The top ten of states with more Mass Killings is showing below:

In [ ]:
MassKillingStates=dataWithBins.groupby(['state']).sum().sort_values('Mass Killing', ascending=False).head(10)
MassKillingStates=MassKillingStates[['Mass Killing']]
MassKillingStates

In [ ]:
numberKillsGroups = pd.get_dummies(dataWithBins['bins'])
numberKillsGroups.head(10)

In [ ]:
workingDataWithDummies=pd.merge(workingData,numberKillsGroups, left_index=True, right_index=True)
workingDataWithDummies.head(10)